### PR TITLE
Script to pull report data

### PR DESCRIPTION
**Pull request by bamader**
## Background
This PR creates a script that pulls all eligible enrollment report data from the historical care 4 kids data site.

## Additional Context
There were definitely some annoyances with figuring out the best way to standardize this.
1. All of the data from 2018 is straight up missing (what's there are just copies of the files from 2019, with the exact same names, but misleadingly listed as 2018)
2. Some of the years of the files are misnumbered (e.g. four months in 2020 have their reports labeled as if they are Jan through March of 2021), which prevents any sort of explicit upper-bounding on pulling files. I used and documented what heuristics I could for pulling the data; some of them might seem a little "magicky" because we can't use numerical time period bounding, but imo that's not a big deal (and also we don't have an alternative).
3. I made the executive decision to pull the last "five" years of data (in quotes because 2018 is gone), which means our period of analysis is 2016-2020. 2015 and earlier have incomplete or inaccessible excel files for their data, and it is _absolutely not worth it_ to use PDF mining to extract tables from PDF files. Four real years of data should be fine for trend projection, especially since that'll be two years before the elimination of one of the registered categories and two years after (I believe Rachel said that happened after 2017, so our missing 2018 creates what's in essence a perfectly split "experiment").

### CODE REVIEW

File: data_sets/historical_care_4_kids/data_pull.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+page = requests.get(BASE_URL)
+soup = BeautifulSoup(page.text, features="lxml")
+
+# Make a directory to put data into, if it doesn't already exist
+if not os.path.exists(DATA_DIR):
+    os.mkdir(DATA_DIR)
+    
+# If the data directory alreasy has files, nothing to do
+existing_files = os.listdir(DATA_DIR)
+if len(existing_files) == 48:
```
Comment (kmagida110) [inline data_sets/historical_care_4_kids/data_pull.py:44]: Can you make this 48 a constant and explain it for future use?

File: data_sets/historical_care_4_kids/data_pull.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+from bs4 import BeautifulSoup
+import requests
+import os
+
+BASE_URL = "https://www.ctcare4kids.com/care-4-kids-program/reports/"
+DATA_DIR = "enrollment_reports/"
+
+'''
```
Comment (kmagida110) [inline data_sets/historical_care_4_kids/data_pull.py:8]: Nit ish but can you put function comments within the function, the standard is that code outside of functions is about the entire system while code inside functions documents the function

File: data_sets/historical_care_4_kids/data_pull.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+care about getting some of the files tagged as 2021.
+'''
+def is_a_report_file(href):
+    # All reports contain either 'files' or 'uploads' in their url string
+    if 'files' in href or 'uploads' in href:
+        # Only want excel versions of files, not PDFs
+        if href[-5:] == '.xlsx':
+            # We care about the last 5 years of data (2016-2020), disregard
+            # earlier years
+            if '2015' not in href:
```
Comment (kmagida110) [inline data_sets/historical_care_4_kids/data_pull.py:28]: Nit, this could be a one line and might read better

**Review by kmagida110 — APPROVED** [review]
A few nits but nothing blocking. I'm 100% OK with your assumptions with time period. I don't think ECIS goes back before 2016 either. The next step here will be converting all of these to a single dataframe/csv. That is not going to be very clean!

**Review by imcurtis — APPROVED** [review]

File: data_sets/historical_care_4_kids/data_pull.py (hunk 3 above)
Comment (bamader) [reply data_sets/historical_care_4_kids/data_pull.py:28]: Yeah I wrote all the `if`'s out as one line and it looked hideous. I found it much easier to figure out what was going on if I separated the conditions to explain my reasoning. I realize it's not the most "optimal" way to write the code, but to me the readability improvement was quite large.

File: data_sets/historical_care_4_kids/data_pull.py (hunk 2 above)
Comment (bamader) [reply data_sets/historical_care_4_kids/data_pull.py:8]: Gonna have to disagree on this one. I have a very strong preference for function docs on the lines right before the function declaration. I think it reads much better and makes it easier for a programmer to know whether they have to read any of the body of a function at all or whether they can skip it. I think moving explanatory comments inside the function also doesn't work as well with the stylistic convention of putting  all the function definitions first in a file, then having script code at the bottom; since the functions are written out prior to code execution, there can't be a risk of confusion  whether something is system level or function level if it's defined in the function part of a code file. Another small benefit to anyone that uses IDEs (myself included) is that if you click the arrows beside functions and you minimize the code body, you can  still scan the docstring to know what the function does if you need to call it without reopening a bunch of functions to check what they do.

As far as it being "standard," idk about this either. In ECE Reporter we have function doctstrings outside the function in every file, and I feel like that's been a good move. We actually at one point had a long semi-random stylistic discussion about that; the analogy that came up was whether comments about conditionals should be  before or after the 'if,'  i.e.

```
// Comment about the test
if (test) { blah }
```

or 

```
if (test) {
    // comment about test
    blah
}
```

We had unanimous preference for the first one because we felt it read cleaner and separated out comments about the overall test from comments about code inside the execution of the if, and I think the logic applies identically to function declarations.

Also, it's funny how fired up we as programmers get over style things like this 😝 . You wrote your comment from the perspective "well of course comments go inside, that's been standard in my experience and I think it reads  better," and literally the second I saw your comment I was like "what sort of foul treason is this, docstrings _clearly_ go outside the function because it's easier to parse and better fits programming structure." Good thing python doesn't do `{}` for `if/else`, otherwise we could argue about the styles of 

```
if () {
    stuff
}
````

vs

```
if ()
{
    stuff
}
````

Anyway, TLDR: I have a strong preference to not put docstring locations inside functions, but if that's a blocker for others then I'll move them. Cuz team first 👥   (I can't make a high-five in a github comment so shadows of a team is the best I can do 🙄 ).

Comment (kmagida110) [reply data_sets/historical_care_4_kids/data_pull.py:14]: I try to avoid holy wars about syntax since I usually really don't care, this is a pretty specific Python/Python IDE exception.  Python itself allows actually calling for the docstring (`help(some_function)` or some_function.__doc__) and Pycharm leverages this as well with help texts. I've also just been linted into submission by Pep/Pycharm and that is a pep standard: https://www.python.org/dev/peps/pep-0257/. I promise not to fight other holy wars, we should get a linter at some point though for general uniformity.
